core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -11,53 +11,240 @@
+        mPageStore.put(fileId, pageIndex, page);
+      } else if (!needEvict) {
+        mEvictor.updateOnPut(fileId, pageIndex);
+        mPageStore.put(fileId, pageIndex, page);
+      }
     }
-    mEvictor.updateOnGet(pageId);
-    return null;
+
+    Pair<ReadWriteLock, ReadWriteLock> pageLockPair =
+        getPageLockPair(fileId, pageIndex, victimFileId, victimPageIndex);
+    try (LockResource r1 = new LockResource(pageLockPair.getFirst().writeLock());
+        LockResource r2 = new LockResource(pageLockPair.getSecond().writeLock())) {
+      try (LockResource r3 = new LockResource(mMetaLock.writeLock())) {
+        if (mMetaStore.hasPage(fileId, pageIndex)) {
+          LOG.warn("fileId {} pageIndex {} is already inserted by a racing thread",
+              fileId, pageIndex);
+          return;
+        }
+        if (!mMetaStore.hasPage(victimFileId, victimPageIndex)) {
+          LOG.warn("fileId {} pageIndex {} is already evicted by a racing thread",
+              fileId, pageIndex);
+          return;
+        }
+        try {
+          mMetaStore.removePage(victimFileId, victimPageIndex);
+        } catch (PageNotFoundException e) {
+          // this should never happen with proper locking
+          LOG.error("failed to remove page {} {} from meta store",
+              victimFileId, victimPageIndex, e);
+        }
+        mEvictor.updateOnDelete(victimFileId, victimPageIndex);
+        mMetaStore.addPage(fileId, pageIndex);
+        mEvictor.updateOnPut(fileId, pageIndex);
+      }
+      try {
+        mPageStore.delete(victimFileId, victimPageIndex);
+      } catch (PageNotFoundException e) {
+        // this should never happen with proper locking
+        LOG.error("failed to delete page {} {} from page store", victimFileId, victimPageIndex, e);
+      }
+      mPageStore.put(fileId, pageIndex, page);
+    }
+  }
+
+  @Override
+  public ReadableByteChannel get(long fileId, long pageIndex) throws IOException,
+      PageNotFoundException {
+    return get(fileId, pageIndex, 0, mPageSize);
   }
 
   @Override
-  public boolean delete(long fileId, long pageId) throws IOException, PageNotFoundException {
-    mMetaStore.removePage(pageId);
-    mPageStore.delete(fileId, pageId);
-    return false;
+  public ReadableByteChannel get(long fileId, long pageIndex, int pageOffset, int length)
+      throws IOException, PageNotFoundException {
+    Preconditions.checkArgument(pageOffset + length <= mPageSize,
+        "Read exceeds page boundary: offset=%s length=%s, size=%s", pageOffset, length, mPageSize);
+    ReadableByteChannel ret;
+    boolean hasPage;
+    ReadWriteLock pageLock = getPageLock(fileId, pageIndex);
+    try (LockResource r = new LockResource(pageLock.readLock())) {
+      try (LockResource r2 = new LockResource(mMetaLock.readLock())) {
+        hasPage = mMetaStore.hasPage(fileId, pageIndex);
+      }
+      if (!hasPage) {
+        throw new PageNotFoundException(
```

Comment:
`return null`. using Exceptions for control flow is undesired.

Reply:
Refactored.